homeassistant/components/knx/storage/config_store.py

```diff
@@ -108,7 +108,7 @@ async def update_entity(
             raise ConfigStoreException(
                 f"Entity not found in storage: {entity_id} - {unique_id}"
             )
-        await self.entities.pop(unique_id).async_remove()
+        async_dispatcher_send(self.hass, SIGNAL_ENTITY_REMOVE.format(unique_id))
         self.async_add_entity[platform](unique_id, data)
```

Comment:
@MartinHjelmare would I need to await the entities `async_remove()` (eg. by passing a future with the dispatcher) before adding the entity again? Or is it safe to do it like this way?

Reply:
You should wait for the entity remove to complete before adding it again.

Reply:
I think it may be better to contain all this code within the platform since then we can access the platform helper and use its methods. Eg, we could get the platform helper with `async_get_current_platform` and then look up the entity instance via the entity_id and operate directly on the entity. Probably define a common function that can be imported in each platform and used.

Reply:
We don't have to use the dispatcher either. The important thing is to split the abstraction correctly between storage and platforms.

Eg, we could create a small class with abstract methods that each platform is responsible to implement where the abstract interface has methods for adding an entity and removing an entity and updating an entity. Then the platforms register this instance to the storage.

Reply:
So like extending what already happens here https://github.com/home-assistant/core/blob/bb31fc1ec72e204ee153f400b9baaa1a71d564e2/homeassistant/components/knx/knx_entity.py#L70

Would you know of any integration that does this already, so I can have a look at an example?

Reply:
So how can this platform helper instance remove an entity (call `Entity.async_remove()`)? Would it also need to dispatch a signal to the entity instance or can it have a reference to the entity itself (like the storage did before)?

Should it be instantiated in the platforms `async_setup_entry()`?

Reply:
It can work on the entity directly since it can access the EntityPlatform instance that holds all the entities of the platform.

Reply:
There's also `EntityPlatform.async_remove_entity`.

Reply:
> Should it be instantiated in the platforms async_setup_entry()?

Yes.

Reply:
Addressed in #123128
Thanks for your suggestions! 👍